File: multiplyColumns/multiplyColumns.go
package main

import (
	"flag"
	"fmt"
	"os"
	"log"
	"runtime/pprof"
)

// profiling
var cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")

var rows [4][][4]int // an array containing a slice which contains an array
var cols [4][][4]int
var board [5][5]int // an array containing an array

func start() {
	board[0][4] = 162
	board[1][4] = 200
	board[2][4] = 147
	board[3][4] = 140
	board[4][0] = 140
	board[4][1] = 150
	board[4][2] = 441
	board[4][3] = 72

	for i := 0; i < 4; i++ {
		rows[i] = possibilities(board[i][4])
		cols[i] = possibilities(board[4][i])
	}
}

func printBoard() {
	for row := 0; row < 5; row++ {
		for col := 0; col < 5; col++ {
			fmt.Printf(" %3d", board[row][col])
		}
		fmt.Println()
	}
}

// could find a better way to get the possibilities I think
func possibilities(prod int) (res [][4]int) {
	for col1 := 1; col1 < 10; col1++ {
		for col2 := 1; col2 < 10; col2++ {
			for col3 := 1; col3 < 10; col3++ {
				for col4 := 1; col4 < 10; col4++ {
					if col1*col2*col3*col4 == prod {
						res = append(res, [4]int{col1, col2, col3, col4})
					}
				}
			}
		}
	}
	return
}

func smartPossibilities(prod int) (res [][4]int) {
	var divisors []int = make([]int, 9)

	divisors = append(divisors, 1)

	for i := 2; i < 10; i++ {
		if prod%i == 0 {
			divisors = append(divisors, i)
		}
	}

	for _, col1 := range divisors {
		for _, col2 := range divisors {
			for _, col3 := range divisors {
				for _, col4 := range divisors {
					if col1*col2*col3*col4 == prod {
						//fmt.Printf("%3d %3d %3d %3d %3d\n", col1, col2, col3, col4, prod)
						res = append(res, [4]int{col1, col2, col3, col4})
					}
				}
			}
		}
	}
	return
}

func shortest(array [4][][4]int) (row int) {
	length := len(array[0])
	for key, value := range array {
		if len(value) < length {
			row = key
			length = len(value)
		}
	}
	return
}

func insert(axis string, index int, array [4]int) {
	if axis == "row" {
		for col := 0; col < 4; col++ {
			board[index][col] = array[col]
		}
	} else {
		for row := 0; row < 4; row++ {
			board[row][index] = array[row]
		}
	}
}

func fillRow(row int) (match bool) {

	// this is kinda hacky
	if row == 4 {
		match = true
		return
	}

	startValue := [4]int{board[row][0], board[row][1], board[row][2], board[row][3]}
	for _, value := range rows[row] {
		match = true
		for col := 0; col < 4 && match; col++ {
			if board[row][col] != 0 && board[row][col] != value[col] {
				match = false
			}
		}
		if match {
			insert("row", row, value)
			if fillCol(row) {
				return
			} else {
				insert("row", row, startValue)
				match = false
			}
		}
	}
	return
}

func fillCol(col int) (match bool) {
	startValue := [4]int{board[0][col], board[1][col], board[2][col], board[3][col]}
	for _, value := range cols[col] {
		match = true
		for row := 0; row < 4 && match; row++ {
			if board[row][col] != 0 && board[row][col] != value[row] {
				match = false
			}
		}
		if match {
			insert("col", col, value)
			if fillRow(col + 1) {
				return
			} else {
				insert("col", col, startValue)
				match = false
			}
		}
	}
	return
}

func main() {

	flag.Parse()
	if *cpuprofile != "" {
		f, err := os.Create(*cpuprofile)
		if err != nil {
			log.Fatal(err)
		}
		pprof.StartCPUProfile(f)
		defer pprof.StopCPUProfile()
	}

	start()
	shortRow := shortest(rows)
	shortCol := shortest(cols)
	found := false

	for row := 0; row < len(rows[shortRow]) && !found; row++ {
		for col := 0; col < len(cols[shortCol]) && !found; col++ {
			if rows[shortRow][row][shortCol] == cols[shortCol][col][shortRow] {
				insert("row", shortRow, rows[shortRow][row])
				insert("col", shortCol, cols[shortCol][col])
				if fillRow(0) {
					printBoard()
					found = true
				}
			}
		}
	}
}

File: multiplyColumns/multiplyColumns.rb
class MultiplyColumns
  def initialize
    @board = [
      [0, 0, 0, 0, 162],
			[0, 0, 0, 0, 200],
			[0, 0, 0, 0, 147],
			[0, 0, 0, 0, 140],
			[140, 150, 441, 72, 0]
    ]
    @rows = []
    @cols = []
    
    4.times do |i|
      @rows[i] = smartPossibilities(@board[i][4])
      @cols[i] = smartPossibilities(@board[4][i])
    end
  end
  
  def printBoard
    5.times do |row|
      5.times do |col|
        print(' %3d' % @board[row][col])
      end
      puts "\n"
    end
  end
  
  def possibilites(prod)
    res = []
    
    for col1 in 1..10
      for col2 in 1..10
        for col3 in 1..10
          for col4 in 1..10
            if (col1 * col2 * col3 * col4) == prod
              res[] = [col1, col2, col3, col4]
            end
          end
        end
      end
    end
    
    res
  end
  
  def smartPossibilities(prod)
    res = []
    divisors = [1]
    
    for i in 2..10
      if (prod % i) == 0
        divisors.push i
      end
    end
    
    divisors.each do |col1|
      divisors.each do |col2|
        divisors.each do |col3|
          divisors.each do |col4|
            if (col1 * col2 * col3 * col4) == prod
              res.push [col1, col2, col3, col4]
            end
          end
        end
      end
    end
    
    res
  end
  
  def shortest(data)
    row = 0
    length = data[0].length
    
    4.times do |i|
      if data[i].length < length
        row = i
        len = data[i].length
      end
    end
    
    row
  end
  
  def insert(axis, index, data)
    if axis == 'row'
      4.times do |col|
        @board[index][col] = data[col]
      end
    else
      4.times do |row|
        @board[row][index] = data[row]
      end
    end
  end
  
  def fillRow(row)
    if row == 4
      return true
    end
    
    startValue = [@board[row][0], @board[row][1], @board[row][2], @board[row][3]]
    match = true
    
    @rows[row].each do |value|
      match = true
      
      4.times do |col|
        if @board[row][col] != 0 && @board[row][col] != value[col]
          match = false
          break
        end
      end
      
      if match
        insert('row', row, value)
        if fillCol(row)
          return match
        else
          insert('row', row, startValue)
          match = false
        end
      end
    end
    
    match
  end
  
  def fillCol(col)
    startValue = [@board[0][col], @board[1][col], @board[2][col], @board[3][col]]
    match = true
    
    @cols[col].each do |value|
      match = true
      4.times do |row|
        if @board[row][col] != 0 && @board[row][col] != value[row]
          match = false
        end
      end
      
      if match
        insert('col', col, value)
        if fillRow(col + 1)
          return match
        else
          insert('col', col, startValue)
          match = false
        end
      end
    end
    
    match
  end
  
  def main
    shortRow = shortest(@rows)
    shortCol = shortest(@cols)
    found = false
    
    @rows[shortRow].each do |row|
      @cols[shortCol].each do |col|
        if row[shortCol] == col[shortRow]
          insert('row', shortRow, row)
          insert('col', shortCol, col)
          
          if fillRow(0)
            printBoard
            return
          end
        end
      end
    end
  end
end

obj = MultiplyColumns.new
obj.main